x-pack/plugins/security_solution/public/management/services/exceptions_list/exceptions_list_api_client.ts

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  CreateExceptionListItemSchema,
+  CreateExceptionListSchema,
+  ExceptionListItemSchema,
+  ExceptionListSummarySchema,
+  FoundExceptionListItemSchema,
+  ListId,
+  UpdateExceptionListItemSchema,
+} from '@kbn/securitysolution-io-ts-list-types';
+import { EXCEPTION_LIST_ITEM_URL, EXCEPTION_LIST_URL } from '@kbn/securitysolution-list-constants';
+import { HttpStart } from 'kibana/public';
+import { MANAGEMENT_DEFAULT_PAGE, MANAGEMENT_DEFAULT_PAGE_SIZE } from '../../common/constants';
+
+/**
+ * A generic class to be used for each artifact type.
+ * It follow the Singleton pattern.
+ * Please, use the getInstance method instead of creating a new instance when using this implementation.
+ */
+export class ExceptionsListApiClient {
+  private static instance: Map<ListId, ExceptionsListApiClient>;
+  private ensureListExists: Promise<void>;
+
+  constructor(
+    private readonly http: HttpStart,
+    private readonly listId: ListId,
+    private readonly listDefinition: CreateExceptionListSchema
+  ) {
+    this.ensureListExists = this.createExceptionList();
```

Comment:
Don't recall if you are done with this PR or not, but this here still has the potential for multiple calls to be done for creating a list. See my suggestion comment below under `createExceptionLIst()`

Also, I would not create this list here. I would suggest creating it only when a method of the class is called. Like that, we are not creating data if the method of a class (for whatever reason) are not called.

Reply:
I'm not sure if I've understood what you mean.

Having this call on the constructor ensures this is only called one time as this class is a Singleton and will generate an instance the first time and will return this instance every time the `getInstance` method is called.

I understand the wondering about having this on the constructor but there is no use case when we instantiate it without making any call and we need to ensure the list exists for all the available calls.

The await on each method just ensures the `createExceptionList` call has finished before running the actions below, but it not generates an extra call, just await the ones generated on the constructor to finish.

Let me know if that make sense or I'm missing anything.

Reply:
The issue I see is that `createExceptionList()` is not keeping track or checking if the list has already been created, so it sends a new request to the API every time.  Here is a code example that maybe clarifies this what I mean (the fact that multiple API calls can still be made:

```typescript
const inst1 = new ExceptionsListApiClient('for trusted apps');

const inst2 = new ExceptionsListApiClient('for trusted apps');

const inst3 = new ExceptionsListApiClient('for trusted apps');
```

I have not checked this out and try to run it myself, but in reading the code, I'm pretty sure 3 API calls to create the list will be sent to the API right? because `ensureListExists` is a class instance member, not a `Class` static instance member. My comment below under the `createExceptionList()` is a possibly solution.

Reply:
I'm good with calling `createExceptionList()` from the constructor and storing the response in the private instance member, but we should ensure the list is not created more than once

Reply:
Why you would call three times the constructor? The `getInstance` method is the one in charge to ensure the constructor is called just one time. There is no restriction on that, but we could add this to the constructor if want to ensure a new instance is created only from the `getInstance` method:

```typescript
  if(ExceptionsListApiClient.instance.has(listId)){
            throw new Error("Error: Instantiation failed: Use ExceptionsListApiClient.getInstance() instead of new.");
        }
```

Reply:
yeah, but that assumes `.getInstance()` would be used all the time. But exposing the class means one would create an instance directly and not request the one that might have already been created and cached.

I think our requirement was: the list should only ever be attempted to be created once

`.getInstance()` is a helper utility and that "flow" for getting a class instance does in fact ensure the list is only created once - not by explicit code, but by caching the class instance. The `class` constructor, however, does not ensure that requirement is met and thus why I think we should address it there (which, btw - does not change the current implementation of `.getInstance()`)

Reply:
done: https://github.com/elastic/kibana/pull/124245/commits/b679c0305c9b7da0b405e767104887bac3eae5c3